# STA130 TUT 01 (Sep06)<br><br> 🏃🏻‍♀️ 🏃🏻 <u> Hitting the ground running... <u><br><br>


In [1]:
1+1

2

In [2]:
print("hello")

hello



## 🚧 🏗️ (Using notebooks and ChatBots) Demo [45 minutes]  
      
#### 1. *[About 8 of the 45 minutes]* Demonstrate going to the course [Quercus homepage](https://q.utoronto.ca/courses/354091); accessing the [Course GitHub Repo](https://github.com/pointOfive/STA130_ChatGPT); opening new and uploaded notebooks on [UofT Jupyterhub](https://datatools.utoronto.ca) (classic jupyter notebook, or jupyterhub is fine, or students may use [google collab](https://colab.research.google.com/)); and using Jupyter notebooks as a "`Python` calculator" and editing ["Markdown cells"](https://www.markdownguide.org/cheat-sheet/)<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> This is all simple and pretty obvious intended to be that way so students get that they can do this on their own.

</details>
    
#### 2. *[About 30 of the 45 minutes]* Demonstrate using [ChatGPT](https://chat.openai.com/) (or [Copilot](https://copilot.microsoft.com/) if conversation privacy is desired) to

1. find an (a) "amusing, funny, or otherwise interesting dataset" which (b) has missing values and is (c) available online through a URL link to a csv file;
2. load the data into the notebook with `pandas` and get "missing data counts" for the dataset;
3. prompt the ChatBot to "Please provide a summary of our interaction for submitting as part of the requirements of an assignment"; and, to "Please provide me with the final working verson of the code that we created"<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> The intention here is to demonstrate that [at the current GPT4.0-ish level] the ChatBot (with probably about an 80% chance) **cannot fullfil all the requests of the inquiry (of (a) "funny or amusing" nature, (b) the presence of missingness, and (c) working url links)** *but will otherwise produce working code*<br><br>
> 
> 1. ChatBots have a notoriously "short term memory"; so, be ready for them to "forget" specific details of your prompting requests 
> 2. ChatBots often cannot pivot away substantially from initial answers; so, be ready for your efforts at follow up and correction with the ChatBot to prove frustratingly futile (which, may in this case actually have a lot to do with the following fact, that...)
> 3. ChatBots don't seem to be very aware of the contents of datasets that are avalable online (or even working url links where datasets are); so, ChatBot are not currently a substitue for exploring dataset repository such as [TidyTuesday](https://github.com/rfordatascience/tidytuesday) (or other data repositiory resources) and reviewing data yourself (although, ChatBot interactions can nonetheless be help with brainstorm dataset ideas and provide a way to "search for content", perhaps especially when referencing a specific website in the conversation)<br><br>
> 
> Examples of this task going pretty well are available [here](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/COP/SLS/00006_copilot_funnyamusingNAdatasetV3.md), [here](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/COP/SLS/00007_copilot_funnyamusingNAdatasetV4.md), and [here](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/GPT/SLS/00001_gpt3p5_villagersdata.md); while, examples of this going poorly are available [here](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/COP/SLS/00002_copilot_funnyamusingNAdataset.md) and [here](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/GPT/SLS/00002_gpt3p5_funnyasusingNAdataset.md). Successes and failures are found within the Microsoft Copilot and ChatGPT ChatBots both, suggesting the quality of the results likely has to do more with "randomness" and perhaps the nature of the prompting and engagement as opposed to the actual ChatBot version being used...
    
</details>

#### 3. *[About 7 of the 45 minutes]* Demonstrate saving your python jupyter notebook in your own account and "repo" on [github.com](https://github.com), and sharing (a) notebook links, (b), ChatBot transcript log links, (c) ChatBot summaries through a piazza post and a Quercus announcement (so students can use this later for their homework assignment if they wish)<br><br>


# First Dataset

Dog Names in NYC Dataset. Chat gave me this

In [3]:
import pandas as pd
url = "https://data.cityofnewyork.us/api/views/25th-nujf/rows.csv"
df = pd.read_csv(url)

Lets look at the head of hour dataset

In [4]:
df.head()

Year of Birth  Gender Ethnicity Child's First Name  Count  Rank
0           2011  FEMALE  HISPANIC          GERALDINE     13    75
1           2011  FEMALE  HISPANIC                GIA     21    67
2           2011  FEMALE  HISPANIC             GIANNA     49    42
3           2011  FEMALE  HISPANIC            GISELLE     38    51
4           2011  FEMALE  HISPANIC              GRACE     36    53

In [5]:
df.Ethnicity

0                  HISPANIC
1                  HISPANIC
2                  HISPANIC
3                  HISPANIC
4                  HISPANIC
                ...        
69209        BLACK NON HISP
69210    WHITE NON HISPANIC
69211    WHITE NON HISPANIC
69212    WHITE NON HISPANIC
69213              HISPANIC
Name: Ethnicity, Length: 69214, dtype: object

This looks fishy, this does not look like dog breeds

# Chat Made a Mistake
## Actual dog names in NYC?

In [6]:
url = "https://data.cityofnewyork.us/api/views/xx67-kt59/rows.csv"
df = pd.read_csv(url)

HTTPError: HTTP Error 404: Not Found

In [7]:
url = "https://data.cityofnewyork.us/api/views/nu7n-tubp/rows.csv"
df = pd.read_csv(url)


/tmp/ipykernel_52/2940949307.py:2: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(url)


In [8]:
df.head()

AnimalName AnimalGender AnimalBirthYear  \
0      PAIGE            F            2014   
1       YOGI            M            2010   
2        ALI            M            2014   
3      QUEEN            F            2013   
4       LOLA            F            2009   

                              BreedName  ZipCode LicenseIssuedDate  \
0  American Pit Bull Mix / Pit Bull Mix  10035.0        09/12/2014   
1                                 Boxer  10465.0        09/12/2014   
2                               Basenji  10013.0        09/12/2014   
3                      Akita Crossbreed  10013.0        09/12/2014   
4                               Maltese  10028.0        09/12/2014   

  LicenseExpiredDate  Extract Year  
0         09/12/2017          2016  
1         10/02/2017          2016  
2         09/12/2019          2016  
3         09/12/2017          2016  
4         10/09/2017          2016

## Missing Values?

In [9]:
# Check for missing values in each column
missing_values = df.isnull().sum()
print(missing_values)

AnimalName            1387
AnimalGender            21
AnimalBirthYear          0
BreedName                0
ZipCode                  9
LicenseIssuedDate        0
LicenseExpiredDate      82
Extract Year             0
dtype: int64


it seems that there are missing values for animal gender and licence expried date 

In [11]:
# Display rows where any value is missing
rows_with_missing_values = df[df.isnull().any(axis=1)]
print(rows_with_missing_values)
print(len(rows_with_missing_values))

       AnimalName AnimalGender AnimalBirthYear  \
1622          NaN            M            2014   
2051          NaN            F            2014   
2843          NaN            M            2014   
3192          NaN            M            2014   
3703          NaN            F            2014   
...           ...          ...             ...   
616757       NONE            F            2023   
616781        NaN            F            2019   
616808        NaN            M            2023   
616828        NaN            M            2022   
616851        NaN            M            2021   

                               BreedName  ZipCode LicenseIssuedDate  \
1622                   Poodle, Miniature  10014.0        01/02/2015   
2051                           Chihuahua  11223.0        01/09/2015   
2843                Schnauzer, Miniature  10032.0        01/20/2015   
3192         Miniature American Shepherd  10016.0        01/25/2015   
3703                         Poodle, Toy  10

# Chat history link

https://chatgpt.com/share/8c621cb8-7249-4c02-b09f-9b59c5985b91

# Summary of the chat 

1. Dataset Request:
You were looking for a fun and interesting dataset that contained missing values and could be easily downloaded and accessed in Python using a direct URL link to a CSV file. You also specified that you didn’t want to use datasets from Kaggle.
2. Initial Suggestions:
I first suggested several datasets (like the Titanic dataset, Air Quality dataset, and more). However, you didn’t find those datasets fun or exciting.
3. Human Names vs. Dog Names Confusion:
I provided a dataset link from NYC’s open data portal, describing it as containing dog names. However, you pointed out that the dataset had an Ethnicity column, which indicated that it was actually about human baby names, not dogs.
I acknowledged the mistake and corrected the suggestion, providing a new, working link for the actual NYC Dog Names dataset.
4. HTTP 404 Error:
You encountered an HTTP Error 404: Not Found when trying to access the corrected dog names dataset. I recognized that the URL was invalid and provided a new working URL that would successfully load the dog names data.
5. DtypeWarning Issue:
After loading the correct dataset, you received a DtypeWarning, indicating that one of the columns had mixed data types (e.g., some rows might have strings, while others might have numbers in the same column).
I explained that this warning occurs when pandas tries to infer data types for large datasets, which can cause issues when there are mixed types in a column.
I then suggested three ways to handle this:
Option 1: Setting low_memory=False, which would process the file in chunks and avoid the warning.
Option 2: Manually specifying the data type of the column to ensure consistency.
Option 3: Ignoring the warning using warnings.filterwarnings() if it doesn't affect your analysis.
6. Checking for Missing Values:
You asked how to check if the dataset had missing values. I provided multiple methods using pandas to check for missing data:
Method 1: Using isnull().sum() to count missing values in each column.
Method 2: Using isnull().any() to check whether any missing values exist in the columns.
Method 3: Using isnull().sum().sum() to get the total count of missing values in the entire dataset.
Method 4: Using df.info() to get a summary of non-null counts in each column.
7. Displaying Rows with Missing Values:
You wanted to see the actual rows that contain missing values. I provided the following code to filter the dataset and display only rows with missing data:
python
Copy code
rows_with_missing_values = df[df.isnull().any(axis=1)]
print(rows_with_missing_values)
This code filters the DataFrame to display rows where at least one value is missing, allowing you to inspect the specific entries that have missing data.
Would you like to explore any of these points further or need help with another task?

## 💬 🗣️ Communication [55 minutes]  
     
#### 1. *[About 15 of the 55 minutes]* Ice breakers  and introductions, in 8 groups of 3 or thereabouts...
    
1. Each person may bring two emojis to a desert island... reveal your emojis at the same time... for emojis selected more than once the group should select one additional emoji
2. Where are you from, what do you think your major might be, and what's an "interesting" fact that you're willing to share about yourself?
        
#### 2. *[About 10 of the 55 minutes]* These are where all the bullet holes were in the planes that returned home so far after some missions in World War II
    
1. Where would you like to add armour to planes for future missions?
2. Hint: there is a hypothetical dataset of the bullet holes on the planes that didn't return which is what we'd ideally compare against the dataset we observe...
        
![Classic image of survivorship bias of WW2 planes](https://upload.wikimedia.org/wikipedia/commons/thumb/b/b2/Survivorship-bias.svg/640px-Survivorship-bias.svg.png)
           
#### 3. *[About 10 of the 55 minutes]* Monte Hall problem: there is a gameshow with three doors, one of which has a prize, and you select one of the doors and the gameshow host reveals one of the other two unchosen doors which does not have the prize... would you like to change your guess to the other unchosen door?

![](https://mathematicalmysteries.org/wp-content/uploads/2021/12/04615-0sxvwbnzvvnhuklug.png)<br>
       
#### 4. *[About 20 of the 60 minutes]* Discuss the experience of the groups for the WW2 planes and Monte Hall problems

1. For each of these problems, have students vote on whether their groups (a) agreed on answers from the beginning, (b) agreed only after some discussion and convincing, or (c) retained somewhat divided in their opinions of the best way to proceed<br><br>
    
2. Briefely identify the correct answer from the answers the groups arrived at<br><br>
    
3. **[If time permits... otherwise this is something students could consider after TUT]** Prompt a [ChatGPT](https://chat.openai.com/) [or [Copilot](https://copilot.microsoft.com/)] ChatBot to introduce and explain "survivorship bias" using spotify songs as an example and see if students are able to generalize this idea for the WW2 planes problem and if they find it to be a convincing argument to understand the problem<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>
    
> This could be done like [this](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/COP/SLS/00009_copilot_survivorshipbias_spotify.md) or [this](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/GPT/SLS/00003_gpt3p5_spotify_Survivorship_Bias.md), or you could instead try to approach things more generally like [this](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/GPT/SLS/00004_gpt3p5_general_Survivorship_Bias.md)
> 
> Two ends of the ChatBot prompting spectrum are
> 
> 1. creating an extensive prompt exhuastively specifying the desired response results; or, 
> 2. iteratively clarifying the desired response results through interactive ChatBot dialogue<br><br>
> 
> This is to some degree a matter of preference regarding the nature of ChatBot conversation sessions, but there it may also be a lever to influence the nature of the responses provided by the ChatBot 
</details>
    
4. **[If time permits... otherwise this is something students could consider after TUT]** Prompt a [ChatGPT](https://chat.openai.com/) [or [Copilot](https://copilot.microsoft.com/)] ChatBot to introduce and explain the Monte Hall problem and see if the students find it understandable and convincing<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>
    
> ChatBots fail to correctly analyze the Monte Hall problem when they're asked for a formal probabilistic argument...
>
> - [ChatGPT fails by wrongly calculating a probability of 1/2...](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/GPT/SLS/00005_gpt3p5_MonteHallWrong.md)
> - [Copilot fares similarly poorly without substantial guidance...](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/COP/SLS/00010_copilot_montehallwrong.md)<br><br>
> 
> *demonstrating (a) that there are clear limits to how deeply ChatBots actually "reason", and (b) that they are instead better understood as simply being information regurgitation machines, and (c) that this means  they can suffer from the "garbage in, garbage out" problem if the quality of the information their responses are based on are is poor and inaccurate (as is notoriously the case in the Monte Hall problem, for which many incorrect mathematical analyses have been "published" into the collection of human generated textual data on which ChatBots are based)*
    
</details>